### DISCUSSION TITLE
Cleaning up Apple Photos Library with Python

### DISCUSSION THREAD

**Discussion by oPromessa** [RhetTbull/osxphotos#752] [Q&A] [answered]
Based on this blog entry [Cleaning up Apple Photos Library with Python](https://www.muchen.ca/blog/2022/cleanup-photos/) is there a way in python using `import osxphotos` library to add pics to an Album in a similar way as the option `--add-exported-to-album "Archived Photos"`does?
```py
album = osxphotos.PhotosAlbum('Bad Photos')
album.add_list(bad_ps)
```

without the _hack_ explained on the blog entry?

"Once we have it installed, we actually need to modify a single line in the library. Go to where the python library is installed, and modify the .../osxphotos/__init__.py file:

Add this line:
```py
from .photosalbum import PhotosAlbum
```
"

**Comment by RhetTbull** [comment, marked as answer]
That hack is not needed. You can simply do this:

```python
from osxphotos.photosalbum import PhotosAlbum
```

Now that I know people are using `PhotosAlbum` I'll expose it in osxphotos. It is an internal class for adding osxphotos `PhotoInfo` objects to an album. If you want to import photos not already in Photos into an album, take a look at [photoscript](https://github.com/RhetTbull/PhotoScript) which is used by `PhotoAlbum`.  If you're using osxphotos to find specific photos as in the blog post and add those to an album, then `PhotosAlbum` is what you want.

**Comment by RhetTbull** [comment]
By the way, you could do what that blog post demonstrates with simply writing a query function.  See [this example](https://github.com/RhetTbull/osxphotos/blob/master/examples/query_function.py).  Assuming you save the query function as `badphotos()` in `badphotos.py` you can use it like this:

`osxphotos query --query-function badphotos.py::badphotos --add-to-album "Bad Photos"`

**Reply by RhetTbull** [reply]
I've added [bad_photos.py](https://github.com/RhetTbull/osxphotos/blob/master/examples/bad_photos.py) to the examples folder -- this can be used as a query function to identify potentially bad photos.

**Reply by oPromessa** [reply]
Super.

For others which may want to adopt the example:

Besides the `not p.screenshot` I'm also considering the following filters: avoid videos; pics classified and documents (don't know if it would be the same as p.screenshot) and not hidden. 

```py
p.isphoto and not p.hidden and "Document" not in p.labels) ]
```

Still testing with the settings but getting the worst pics with the following criteria:


```py
p.score.failure < -0.5,
p.score.overall < 0.1
```

**Reply by RhetTbull** [reply]
The label names might be localized so that might not work for everyone but you could specify it on the command line as a query eval:

`--query-eval "'Document' not in photo.labels"`

If you get the parameters tuned submit a PR or post the tuned parameters here so I can update the example. It makes sense that `score.overall` might give good results as I assume that's a weighted combination of the other scores.

**Reply by RhetTbull** [reply]
Also, you can experiment with only the command line if you're only tweaking a couple of score parameters:

`osxphotos query --query-eval "photo.score.overall < 0.1" --query-eval "photo.score.failure < -0.5" --query-eval "'Document' not in photo.labels"  --only-photos --not-screenshot --add-to-album "Bad Photos"`
